### PR TITLE
docs: add changelog conventions to contribution guidelines

### PR DESCRIPTION
**Pull request by glennsl**

### CODE REVIEW

File: docs/docs/for-developers/contributing.md [hunk 1]
```diff
@@ -24,6 +24,65 @@ To improve the chances to get a pull request merged, you should select an issue
 
 In addition, Onivim 2 is built on [Revery](https://github.com/revery-ui/revery) - any work or improvements there will directly improve Onivim 2, as well!
 
+#### Changelog generation
+
+We use a script to generate the change log based on metadata picked up from the commit message, PR title and PR body. For the changelog to be generated successfully it's therefore important that these follow the exact format explained below.
```
Comment (bryphe) [inline docs/docs/for-developers/contributing.md:29]: Looks great, @glennsl - thanks for adding these details!

File: docs/docs/for-developers/contributing.md [hunk 2]
```diff
@@ -24,6 +24,65 @@ To improve the chances to get a pull request merged, you should select an issue
...
+* **chore:** Changes to the build process or auxiliary tools and libraries such as documentation generation
+* **style:** Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
+
+`scope` is optional, but "encouraged". It is typically one of the "area" [issue labels](https://github.com/onivim/oni2/labels) or the name of a feature project converted to "humanized" title case. E.g. instead of `language-support` use `Language Support`, i.e. two words with the first letter of each word capitalized.
+
+`subject` should contain a succinct description of the change:
+
+* use the imperative, present tense: "change" not "changed" nor "changes"
+* don't capitalize first letter
+* no dot (.) at the end
```
Comment (bryphe) [inline docs/docs/for-developers/contributing.md:56]: Thanks for calling these out! Nice to have specific criteria for this.

**Comment by CrossR** [comment]
I've not checked so maybe this was added external to this PR, but are we linking to this in the Pull Request template? Would aid discoverability for first time contributors.

https://help.github.com/en/github/building-a-strong-community/about-issue-and-pull-request-templates#pull-request-templates

(Its a tradeoff, since it will add it for us all, so its potentially annoying to users who know, but still).

**Comment by glennsl** [comment]
If we add a CONTRIBUTING file, first-time contributors will also get a popup asking them to review the guidelines. Less visible, but also less invasive and annoying.

We're going to review this anyway, so it's not a huge issue if they get it wrong.

File: docs/docs/for-developers/contributing.md [hunk 3]
```diff
@@ -24,6 +24,65 @@ To improve the chances to get a pull request merged, you should select an issue
...
+* **feat:** A new feature
+* **fix:** A bug fix
+* **refactor:** A code change that neither fixes a bug nor adds a feature
+* **perf:** A code change that improves performance
+* **test:** Adding missing or correcting existing tests
+* **docs:** Documentation only changes
+* **chore:** Changes to the build process or auxiliary tools and libraries such as documentation generation
+* **style:** Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
+
+`scope` is optional, but "encouraged". It is typically one of the "area" [issue labels](https://github.com/onivim/oni2/labels) or the name of a feature project converted to "humanized" title case. E.g. instead of `language-support` use `Language Support`, i.e. two words with the first letter of each word capitalized.
```
Comment (glennsl) [inline docs/docs/for-developers/contributing.md:50]:
```suggestion
`scope` is optional, but "encouraged". It is typically one of the "area" [issue labels](https://github.com/onivim/oni2/labels) or the name of a feature project.
```

I noticed you got this wrong on your first attempt @bryphe, so maybe we should just automatically convert `language-support`, `language support` and `LanguageSupport` to `Language Support` instead :wink:

Comment (bryphe) [reply docs/docs/for-developers/contributing.md:50]: oops 😄 Yes, perhaps automatically changing it to 'capitalized human case' would be easiest

**Comment by glennsl** [comment]
A couple "edge cases" I've found looking through recent commits:

- Dependency commits don't feat neatly into the current types. Broadly speaking I'd say it's a `chore`, but it doesn't really fit the description, and its purpose is usually to add a feature or fix a bug. We could add another type for `dependency`, but perhaps these just shouldn't be stand-alone, or should be described differently. Examples: #1690, #1700, #1712, #1714

- Bugfixes often fix existing issues, and should reference those, but there's currently no convention for doing so. I propose `fix(<scope>/#<issue>): <subject>`. Perhaps the `/` could be a more general convention for adding refinement or supplementary information, as sometimes feature commits close issues too. Examples: #1694, #1705, #1713. Examples of features closing issues: #1703, #1717

Also, I wonder if we should move to using labels for temporary "tags" like `[WIP]` and `[Experimental]` instead of putting them in the title. Or just use drafts for `[WIP]`. That would distinguish them better and make the title less noisy now that we're stuffing even more information into it. And I know I've committed PRs with those tags still in them.

File: docs/docs/for-developers/contributing.md [hunk 4]
```diff
@@ -24,6 +24,65 @@ To improve the chances to get a pull request merged, you should select an issue
...
+
+* **feat:** A new feature
+* **fix:** A bug fix
+* **refactor:** A code change that neither fixes a bug nor adds a feature
+* **perf:** A code change that improves performance
+* **test:** Adding missing or correcting existing tests
+* **docs:** Documentation only changes
+* **chore:** Changes to the build process or auxiliary tools and libraries such as documentation generation
+* **style:** Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
+
```
Comment (glennsl) [inline docs/docs/for-developers/contributing.md:49]:
```suggestion
* **docs:** Documentation only changes
* **chore:** Changes to the build process or auxiliary tools and libraries such as documentation generation

```

I don't think this is needed, as we check this in CI anyway. And if for some reason we do on occasion get something like this, we can just do it as a chore I think.

**Comment by glennsl** [comment]
Should we change the branch naming conventions to be based on this as well?

Basically:
```
<type>/<scope>/<summary>
```

**Comment by bryphe** [comment]
> Should we change the branch naming conventions to be based on this as well?

Sounds good to me!

**Pull request merged**